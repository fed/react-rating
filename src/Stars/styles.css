.Star {
  display: inline-block;
}

.Star__radio {
  border: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.Star__label {
  display: inline-block;
  height: 25px;
  padding: 0 0 0 37px;
  position: relative;
  width: 0;
}

.Star:last-of-type .Star__label {
  padding-left: 28px;
}

.Star__radio + .Star__label::before {
  left: 0;
  position: absolute;
  top: 0;
  width: 28px;
}

.Star__radio + .Star__label::before {
  content: url("star-inactive.svg");
}

.Star--active .Star__label::before,
.Star__radio:enabled:hover + .Star__label::before,
.Star__radio:checked + .Star__label::before,
.Star__radio:focus + .Star__label::before,
.Star__radio:checked:focus + .Star__label::before {
  content: url("star-active.svg");
}

.Star__radio:hover + .Star__label::before {
  cursor: pointer;
}

/*
 * Once the form is submitted, the user should
 * no longer be able to interact with the it.
 */
.Star__radio:disabled:hover + .Star__label::before {
  cursor: default;
}
