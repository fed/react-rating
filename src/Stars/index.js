import React from 'react';
import PropTypes from 'prop-types';
import uniqueId from 'lodash/uniqueId';
import './styles.css';

export default function Stars({
  selected,
  total = 5,
  disabled = false,
  className = '',
  onChange
}) {
  // We generate a unique random ID so that we can have multiple
  // instances of the component on the same page.
  const id = uniqueId('rating-');
  const stars = Array(total)
    .fill(0)
    .map((el, index) => {
      const value = index + 1;
      const name = `${id}-stars-${value}`;

      return (
        <span key={value} className="Star">
          <input
            type="radio"
            name={`stars-${id}`}
            value={value}
            id={name}
            className="Star__radio"
            onChange={onChange}
            checked={selected === value}
            disabled={disabled} />

          <label
            className="Star__label"
            htmlFor={name} />
        </span>
      );
    });

  return (
    <fieldset className={className}>
      {stars}
    </fieldset>
  );
}

Stars.propTypes = {
  className: PropTypes.string,
  total: PropTypes.number,
  selected: PropTypes.number,
  disabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired
};
