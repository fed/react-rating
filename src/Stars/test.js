import React from 'react';
import renderer from 'react-test-renderer';
import {mount, shallow} from 'enzyme';
import Stars from '.';
import noop from 'lodash/noop';

// We need to mock Lodash's uniqueId generator so that we can generate
// a valid snapshot and query the component.
jest.mock('lodash/uniqueId', () => jest.fn(() => 'rating-some-random-id'));

describe('Stars component', () => {
  it('renders without crashing', () => {
    mount(<Stars onChange={noop} />);
  });

  it('renders correctly with the default values', () => {
    const tree = renderer.create(
      <Stars onChange={noop} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders any number of stars', () => {
    const wrapper = shallow(<Stars onChange={noop} total={3} />);

    expect(wrapper.children()).toHaveLength(3);
  });
});
