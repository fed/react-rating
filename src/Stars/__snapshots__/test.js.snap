// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stars component renders correctly with the default values 1`] = `
<fieldset
  className=""
>
  <span
    className="Star"
  >
    <input
      checked={false}
      className="Star__radio"
      disabled={false}
      id="rating-some-random-id-stars-1"
      name="stars-rating-some-random-id"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="Star__label"
      htmlFor="rating-some-random-id-stars-1"
    />
  </span>
  <span
    className="Star"
  >
    <input
      checked={false}
      className="Star__radio"
      disabled={false}
      id="rating-some-random-id-stars-2"
      name="stars-rating-some-random-id"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="Star__label"
      htmlFor="rating-some-random-id-stars-2"
    />
  </span>
  <span
    className="Star"
  >
    <input
      checked={false}
      className="Star__radio"
      disabled={false}
      id="rating-some-random-id-stars-3"
      name="stars-rating-some-random-id"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="Star__label"
      htmlFor="rating-some-random-id-stars-3"
    />
  </span>
  <span
    className="Star"
  >
    <input
      checked={false}
      className="Star__radio"
      disabled={false}
      id="rating-some-random-id-stars-4"
      name="stars-rating-some-random-id"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="Star__label"
      htmlFor="rating-some-random-id-stars-4"
    />
  </span>
  <span
    className="Star"
  >
    <input
      checked={false}
      className="Star__radio"
      disabled={false}
      id="rating-some-random-id-stars-5"
      name="stars-rating-some-random-id"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="Star__label"
      htmlFor="rating-some-random-id-stars-5"
    />
  </span>
</fieldset>
`;
